feat: exec policy integration in shell mcp

zhao-oai opened the pull request:
adding execpolicy support into the `posix` mcp

bolinfest commented on codex-rs/exec-server/src/posix.rs:
```diff
@@ -146,26 +169,63 @@ pub async fn main_execve_wrapper() -> anyhow::Result<()> {
+    let command: Vec<String> = std::iter::once(format_program_name(file, strip_program_paths))
+        .chain(argv.iter().cloned())
+        .collect();
+    let commands = parse_shell_lc_plain_commands(&command).unwrap_or_else(|| vec![command]);
```
I don't think we should be doing the bash -lc stripping here.

zhao-oai replied:
fixed. as discussed in slack, we should not be supporting chained commands within an individual posix mcp call

bolinfest commented on codex-rs/exec-server/src/posix.rs:
```diff
@@ -146,26 +169,63 @@ pub async fn main_execve_wrapper() -> anyhow::Result<()> {
+
+async fn load_exec_policy() -> anyhow::Result<Arc<Policy>> {
+    let codex_home = find_codex_home().context("failed to resolve codex_home for execpolicy")?;
+    let policy_dir = codex_home.join(POLICY_DIR_NAME);
```
I would rather not copy this `"policy"` constant into this file, especially if it's just for a debug statement.

Why shouldn't the debug statement be in exec_policy_for?

zhao-oai commented on codex-rs/exec-server/src/posix.rs:
```diff
@@ -146,26 +169,63 @@ pub async fn main_execve_wrapper() -> anyhow::Result<()> {
+    strip_program_paths: bool,
+) -> ExecPolicyOutcome {
+    let command: Vec<String> = std::iter::once(format_program_name(file, strip_program_paths))
+        .chain(argv.iter().cloned())
```
done -- skipping `argv[0]` since that also contains program path

bolinfest commented on codex-rs/exec-server/src/posix.rs:
```diff
@@ -146,26 +167,62 @@ pub async fn main_execve_wrapper() -> anyhow::Result<()> {
-        ExecPolicyOutcome::Allow {
-            run_with_escalated_permissions: false,
-        }
+        path.display().to_string()
```
This also applies here: instead of doing `display()`, just do https://doc.rust-lang.org/std/path/struct.Path.html#method.to_str which returns `Option`.

bolinfest committed:
chore: refactor to move Arc<RwLock> concern outside exec_policy_for (#7615)

The caller should decide whether wrapping the policy in `Arc<RwLock>` is
necessary. This should make https://github.com/openai/codex/pull/7609 a
bit smoother.

- `exec_policy_for()` -> `load_exec_policy_for_features()`
- introduce `load_exec_policy()` that does not take `Features` as an arg
- both return `Result<Policy, ExecPolicyError>` instead of
Result<Arc<RwLock<Policy>>, ExecPolicyError>`

This simplifies the tests as they have no need for `Arc<RwLock>`.

.

zhao-oai commented on codex-rs/exec-server/src/posix.rs:
```diff
@@ -146,26 +167,62 @@ pub async fn main_execve_wrapper() -> anyhow::Result<()> {
+        path.file_name()
+            .and_then(|name| name.to_str())
+            .map(str::to_string)
+            .unwrap_or_else(|| path.display().to_string())
```
`evaluate_exec_policy` now returns a `Result<EscalateAction, rmcp::ErrorData>`

bolinfest commented on codex-rs/exec-server/src/posix.rs:
```diff
@@ -87,6 +98,10 @@ struct McpServerCli {
     bash_path: Option<PathBuf>,
+
+    /// Strip program paths before applying execpolicy (e.g., /usr/bin/echo -> echo).
+    #[arg(long, default_value_t = true)]
```
Can you rename this so it makes sense where the default value is `false`?

bolinfest commented on codex-rs/exec-server/src/posix.rs:
```diff
@@ -146,26 +162,63 @@ pub async fn main_execve_wrapper() -> anyhow::Result<()> {
+    let policy = load_exec_policy_for_features(&Features::with_defaults(), &codex_home)
+        .await
+        .map_err(|err: ExecPolicyError| anyhow::anyhow!(err))?;
+    Ok(Arc::new(policy))
```
In general, let the caller decide whether they want to wrap it in `Arc` or not.

zhao-oai replied:
done -- moved the arc wrapping to the caller

bolinfest commented on codex-rs/exec-server/src/posix/mcp_escalation_policy.rs:
```diff
@@ -12,14 +14,10 @@ use crate::posix::escalate_protocol::EscalateAction;
-/// command.
-pub(crate) type ExecPolicy = fn(file: &Path, argv: &[String], workdir: &Path) -> ExecPolicyOutcome;
+/// In-memory execpolicy rules that drive how to handle an exec() call.
+pub(crate) type ExecPolicy = Arc<Policy>;
```
I don't think we need or want this type alias anymore.

Also, do we need to `Arc` it?

bolinfest reviewed (approved):
Let's get this in and I can change, as necessary.
